### DISCUSSION TITLE
Regular expression produces different results depending whether it's stored in a variable or not

### DISCUSSION THREAD

**Discussion by romfrolov** [nodejs/help#5043] [Q&A] [answered]
### Version

v16.13.2

### Platform

Darwin Romans-MBP.zyxel.box 22.3.0 Darwin Kernel Version 22.3.0: Thu Jan  5 20:49:43 PST 2023; root:xnu-8792.81.2~2/RELEASE_ARM64_T8103 arm64

### What steps will reproduce the bug?

1. Launch Node.js CLI

```bash
node
```

2. Test the following 2 paths with the same regular expression

```js
> /\/teams(?:\/)?/g.test('/teams')
true
> /\/teams(?:\/)?/g.test('/teams/')
true
```

3. Declare a variable

```js
> const re = /\/teams(?:\/)?/g;
undefined
```

4. Repeat the test

```js
> re.test('/teams')
true
> re.test('/teams/')
false
```

### How often does it reproduce? Is there a required condition?

Always.

### What is the expected behavior?

Same result regardless whether the regular expression is stored in a variable or not.

### What do you see instead?

Different results.

### Additional information

I was also able to reproduce it in Brave console (Version 1.48.158 Chromium: 110.0.5481.77 (Official Build) (arm64)).

**Comment by cjihrig** [comment, marked as answer]
I've converted this to a discussion because it's not a bug. The `g` flag on the regular expression maintains state. When you use a regular expression literal, you are creating a new regular expression each time. When you assign it to a variable, you are reusing the same regular expression object, so the state of the `g` flag becomes relevant.

You can read more here: https://javascript.plainenglish.io/most-surprising-behavior-of-javascript-regular-expression-you-have-ever-seen-1ddb84539163
